core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -111,19 +112,28 @@ public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionD
                         TopicIdPartition topicIdPartition = data._1;
                         FetchPartitionData fetchPartitionData = data._2;
 
-                        ShareFetchResponseData.PartitionData partitionData = new ShareFetchResponseData.PartitionData()
-                            .setPartitionIndex(topicIdPartition.partition())
-                            .setRecords(fetchPartitionData.records)
-                            .setErrorCode(fetchPartitionData.error.code())
-//                        .setAcquiredRecords(fetchPartitionData.records)
-                            .setAcknowledgeErrorCode(Errors.NONE.code());
-
-                        SharePartition sharePartition = partitionCacheMap.get(sharePartitionKey(groupId, topicIdPartition));
-                        sharePartition.update(memberId, fetchPartitionData);
-
-                        result.put(topicIdPartition, partitionData);
-                        future.complete(result);
+                        partitionCacheMap.get(sharePartitionKey(groupId, topicIdPartition))
+                            .acquire(memberId, fetchPartitionData).whenComplete((acquiredRecords, throwable) -> {
+                                ShareFetchResponseData.PartitionData partitionData = new ShareFetchResponseData.PartitionData()
+                                    .setPartitionIndex(topicIdPartition.partition());
+
+                                if (throwable != null) {
+                                    partitionData.setErrorCode(Errors.forException(throwable).code());
+                                } else {
+                                    // Maybe check if no records are acquired and we want to retry replica
+                                    // manager fetch. Depends on the share partition manager implementation,
+                                    // if we want parallel requests for the same share partition or not.
+                                    partitionData
+                                        .setPartitionIndex(topicIdPartition.partition())
+                                        .setRecords(fetchPartitionData.records)
+                                        .setErrorCode(fetchPartitionData.error.code())
+                                        .setAcquiredRecords(acquiredRecords)
+                                        .setAcknowledgeErrorCode(Errors.NONE.code());
+                                }
+                                result.put(topicIdPartition, partitionData);
+                            });
                     });
+                    future.complete(result);
```

Comment:
Shouldn't the future complete when the acquire completes?

Reply:
True, that's what happens. If there is no overlap then the method here acquires the complete batch from the request. I have changed the code to return directly from here to make it more readable.